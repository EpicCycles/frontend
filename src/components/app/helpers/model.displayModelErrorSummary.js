import {customerAddressFields} from "../model/helpers/fields";
import {displayModelErrorSummary} from "../model/helpers/model";

test('if no error or error_detail undefined returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6
    };
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(undefined);
});
test('if just error no error_detail error returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error: "Model level error"
    };
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe("Model level error");
});
test('if no error with error_detail error_details formatted returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error_detail: {
            "address1": ["A valid address is required.", "SHow me the address is required."],
            "postcode": ["This field may not be blank."]
        },
    };
    const expectedText = "Address: A valid address is required. SHow me the address is required.<br>Postcode: This field may not be blank.";
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(expectedText);
});
test('if error and error_detail formatted returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error: "Model level error",
        error_detail: {
            "address1": ["A valid address is required.", "SHow me the address is required."],
            "postcode": ["This field may not be blank."]
        },
    };
    const expectedText = "Model level error<br>Address: A valid address is required. SHow me the address is required.<br>Postcode: This field may not be blank.";
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(expectedText);
});