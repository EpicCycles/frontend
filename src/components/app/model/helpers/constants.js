export const POSTCODE_RULES = [
  { countryCode: 'AT', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'AU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'BE', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'BG', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CA', regex: /^[A-Za-z][0-9][A-Za-z] [0-9][A-Za-z][0-9]/, display: 'ANA NAN' },
  { countryCode: 'CH', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CY', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CZ', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'DE', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'DK', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'EE', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'ES', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'FI', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'FR', regex: /^\d{5}$/, display: 'NNNNN' },
  {
    countryCode: 'GB',
    regex: /^[A-Z](([0-9][0-9]?)|([A-Z][0-9][0-9]?)|([A-Z]?[0-9][A-Z])) ?[0-9][ABD-HJLNP-UW-Z]{2}$/,
    display: 'CCNN NCC',
  },
  { countryCode: 'GR', regex: /^\d{3}\s\d{2}$/, display: 'NNNNN or NNN NN' },
  { countryCode: 'HU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'IT', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'KR', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'LT', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'LU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'MT', regex: /^\[A-Z]{3}\s\d{4}$/, display: 'AAA NNNN' },
  { countryCode: 'MY', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'NO', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'NZ', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'PH', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'PL', regex: /^\d{2}-\d{3}$/, display: 'NN-NNN' },
  { countryCode: 'PT', regex: /^\d{4}-\d{3}$/, display: 'NNNN-NNN' },
  { countryCode: 'RO', regex: /^\d{6}$/, display: 'NNNNNN' },
  { countryCode: 'SE', regex: /^\d{3,4}$/, display: 'NNNNN or NNNN' },
  { countryCode: 'SG', regex: /^\d{6}$/, display: 'NNNNNN' },
  { countryCode: 'SI', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'SK', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'US', regex: /^(\d{5}([-]\d{4})?)$/, display: 'NNNNN-NNNN or NNNNN' },
  { countryCode: 'ZA', regex: /^\d{4}$/, display: 'NNNN' },
];
