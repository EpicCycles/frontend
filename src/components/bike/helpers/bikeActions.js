export const bikeActions = (
  bikeId,
  bikeReviewList,
  bikeDeleteList,
  changeBikeReviewList,
  changeBikeDeleteList,
) => {
  const onReviewList = bikeReviewList.includes(bikeId);
  const onDeleteList = bikeDeleteList.includes(bikeId);
  return [
    {
      iconName: 'edit outline',
      iconTitle: 'Review this bike',
      className: onReviewList ? 'red' : '',
      iconAction: () => !onDeleteList && changeBikeReviewList(bikeId),
      iconDisabled: onDeleteList,
    },
    {
      iconName: 'delete',
      iconTitle: 'Delete this bike',
      className: onDeleteList ? 'red' : '',
      iconAction: () => !onReviewList && changeBikeDeleteList(bikeId),
      iconDisabled: onReviewList,
    },
  ];
};
