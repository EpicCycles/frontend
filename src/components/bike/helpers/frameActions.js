export const frameActions = (
  frameId,
  frameArchiveList,
  frameDeleteList,
  changeFrameArchiveList,
  changeFrameDeleteList,
) => {
  const onArchiveList = frameArchiveList.includes(frameId);
  const onDeleteList = frameDeleteList.includes(frameId);
  return [
    {
      iconName: 'archive',
      iconTitle: 'Archive this frame and all related bikes',
      className: onArchiveList ? 'red' : '',
      iconAction: () => !onDeleteList && changeFrameArchiveList(frameId),
      iconDisabled: onDeleteList,
    },
    {
      iconName: 'delete',
      iconTitle: 'Delete this frame and all related bikes',
      className: onDeleteList ? 'red' : '',
      iconAction: () => !onArchiveList && changeFrameDeleteList(frameId),
      iconDisabled: onArchiveList,
    },
  ];
};
