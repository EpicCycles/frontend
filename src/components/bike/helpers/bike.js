import {buildPartObject} from "../../part/helpers/part";
import {buildBrandNameArray, getBrandName} from "../../brand/helpers/brand";
import {findObjectWithId} from "../../../helpers/utils";

export const buildDataForApi = (brand, frameName, rowMappings, uploadedHeaders, uploadedData, brands) => {
    const brandsLower = buildBrandNameArray(brands);

    const frame = {
        brand: brand,
        frame_name: frameName
    };
    // start by building an array of bike objects.
    const bikeNames = uploadedHeaders.slice(1);
    let bikes = bikeNames.map(bikeName => {
        return { model_name: bikeName, parts: [] };
    });
    const numberOfBikes = bikes.length;

    rowMappings.forEach(rowMapping => {
        if (!rowMapping.ignore) {
            const dataToUse = uploadedData[rowMapping.rowIndex].slice(1);
            dataToUse.forEach((dataValue, bikeIndex) => {
                if ((bikeIndex < numberOfBikes) && (dataValue.trim().length > 0)) {
                    if (rowMapping.bikeAttribute) {
                        bikes[bikeIndex][rowMapping.bikeAttribute] = dataValue;
                    } else if (rowMapping.partType) {
                        bikes[bikeIndex].parts.push(buildPartObject(
                            rowMapping.partType,
                            dataValue,
                            brandsLower,
                            brand
                        ));
                    }
                }
            });
        }
    });
    frame.bikes = bikes;
    return frame;
};

export const bikeFullName = (bike, frames, brands) => {
    if (bike.frame_name) return `${bike.frame_name} ${bike.model_name}`;
    const frame = findObjectWithId(frames, bike.frame);
    if (!frame) return `${bike.model_name} (unknown brand and frame)`;
    return `${getBrandName(frame.brand, brands)}: ${frame.frame_name} ${bike.model_name}`
};

export const getBikeName = (bikeId, bikes, frames, brands) => {
        const bike = findObjectWithId(bikes, bikeId);
        if (bike) return bikeFullName(bike, frames, brands);
        return 'Unknown Bike'
};

export const findPartsForBike = (bike, bikeParts, parts) => {
    if (! (bike && bikeParts && parts)) return [];
    return findPartsForBikeId(bike.id, bikeParts, parts);
};
export const findPartsForBikeId = (bikeId, bikeParts, parts) => {
    if (! (bikeId && bikeParts && parts)) return [];
    return bikeParts.filter(bikePart => bikePart.bike === bikeId).map(bikePart => {
        return findObjectWithId(parts, bikePart.part)
    });
};
