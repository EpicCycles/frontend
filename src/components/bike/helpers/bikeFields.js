import {
  CLUB_PRICE_FIELD,
  COLOURS_FIELD,
  DESCRIPTION_FIELD,
  EPIC_PRICE_FIELD,
  MODEL_NAME_FIELD,
  PART_TYPE_FIELD,
  PRODUCT_CODE_FIELD,
  RRP_FIELD,
  SIZES_FIELD,
  TEXT,
} from '../../app/model/helpers/fields';
import { PART_NAME_MISSING } from '../../app/model/helpers/error';

export const bikeFields = [
  MODEL_NAME_FIELD,
  DESCRIPTION_FIELD,
  PRODUCT_CODE_FIELD,
  COLOURS_FIELD,
  RRP_FIELD,
  EPIC_PRICE_FIELD,
  CLUB_PRICE_FIELD,
  SIZES_FIELD,
];
export const bikePartFields = [
  PART_TYPE_FIELD,
  {
    fieldName: 'partName',
    header: 'Part Name',
    required: true,
    error: PART_NAME_MISSING,
    type: TEXT,
    displaySize: 40,
    maxLength: 200,
    maxWidth: '250px',
  },
];
