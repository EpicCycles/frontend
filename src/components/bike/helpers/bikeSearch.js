import { ARCHIVED, BRAND, CHECKBOX, TEXT } from '../../app/model/helpers/fields';
export const REQUIRE_VALUES = 'Search criteria must be entered';
export const bikeSearchData = (updatedValue, fullModelData) => {
  if (!fullFields.some(field => field.searchField && !!fullModelData[field.fieldName]))
    return REQUIRE_VALUES;
};
const brandField = {
  fieldName: BRAND,
  header: 'Brand: ',
  displayName: 'Brand:',
  type: BRAND,
  validator: bikeSearchData,
  searchField: true,
  bikeOnly: true,
};
const frameNameField = {
  validator: bikeSearchData,
  displayName: 'Frame Name like:',
  header: 'Frame Name like:',
  fieldName: 'frameName',
  searchField: true,
  type: TEXT,
};
const archivedSelectField = {
  header: 'Include archived Frames:',
  displayName: 'Include archived Frames:',
  fieldName: ARCHIVED,
  type: CHECKBOX,
};
const fullFields = [brandField, frameNameField, archivedSelectField];
const partialFields = [brandField, frameNameField, archivedSelectField];

export const bikeSearchFields = canSelectArchived => {
  if (canSelectArchived) return fullFields;
  return partialFields;
};
