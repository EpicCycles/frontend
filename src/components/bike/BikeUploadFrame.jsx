import BrandSelect from '../brand/BrandSelect';
import FormTextInput from '../../common/FormTextInput';
import React, { Fragment } from 'react';
import BrandModal from '../brand/BrandModal';
import { NEW_ELEMENT_ID } from '../../helpers/constants';

class BikeUploadFrame extends React.Component {
  constructor() {
    super();
    this.state = {
      showModal: false,
    };
    this.handleOpenModal = this.handleOpenModal.bind(this);
    this.handleCloseModal = this.handleCloseModal.bind(this);
  }

  handleOpenModal() {
    this.setState({ showModal: true });
  }

  handleCloseModal() {
    this.setState({ showModal: false });
  }

  saveBrand = brand => {
    const brandsWithUpdates = this.props.brands.slice();
    brandsWithUpdates.push(brand);
    this.props.saveBrands(brandsWithUpdates);
  };

  render() {
    const { showModal } = this.state;
    const {
      brands,
      suppliers,
      onChange,
      brandSelected,
      displayOnly,
      frameName,
      brandName,
    } = this.props;
    return (
      <section className="row">
        {showModal && (
          <BrandModal
            brandModalOpen={showModal}
            componentKey={NEW_ELEMENT_ID}
            saveBrand={this.saveBrand}
            closeBrandModal={this.handleCloseModal}
            suppliers={suppliers}
          />
        )}
        <div key="bikeUpload" className="grid">
          <div className="grid-row">
            <div className="grid-item--borderless field-label">Bike Brand</div>
            <div className="grid-item--borderless">
              {displayOnly ? (
                <Fragment>{brandName}</Fragment>
              ) : (
                <Fragment>
                  <BrandSelect
                    brands={brands}
                    fieldName="brand"
                    onChange={onChange}
                    brandSelected={brandSelected}
                    isEmptyAllowed={true}
                    bikeOnly={true}
                  />
                  <button onClick={this.handleOpenModal}>Add Brand</button>
                </Fragment>
              )}
            </div>
          </div>

          <div className="grid-row">
            <div className="grid-item--borderless field-label">Frame Name</div>
            <div className="grid-item--borderless">
              {displayOnly ? (
                <Fragment>{frameName}</Fragment>
              ) : (
                <FormTextInput
                  id="frameName"
                  fieldName="frameName"
                  placeholder="Frame Name"
                  value={frameName}
                  onChange={onChange}
                  size={100}
                />
              )}
            </div>
          </div>
        </div>
      </section>
    );
  }
}

export default BikeUploadFrame;
