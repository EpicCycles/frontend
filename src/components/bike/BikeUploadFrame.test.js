import BikeUploadFrame from './BikeUploadFrame';
import { NEW_ELEMENT_ID } from '../../helpers/constants';

const brands = [
  { id: 1, brand_name: 'brand 1' },
  { id: 2, brand_name: 'brand 2', supplier: [] },
  { id: 3, brand_name: 'brand 3', supplier: [1] },
  { id: 4, brand_name: 'brand 4', delete: true },
  {
    id: 5,
    brand_name: 'brand 5',
    changed: true,
    supplier: [1, 3],
  },
  { dummyKey: '123ABC', brand_name: 'brand new', changed: true },
  { dummyKey: NEW_ELEMENT_ID, brand_name: 'brand new 2', changed: true },
];
test('upload frame details displays when none entered', () => {
  const component = shallow(<BikeUploadFrame brands={brands} />);
  expect(toJson(component)).toMatchSnapshot();
});
test('upload frame details displays when frame entered', () => {
  const component = shallow(
    <BikeUploadFrame brands={brands} brandSelected={brands[0].id} frameName="bicicletta" />,
  );
  expect(toJson(component)).toMatchSnapshot();
});
