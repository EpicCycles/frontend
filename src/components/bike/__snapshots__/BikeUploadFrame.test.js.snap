// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload frame details displays when frame entered 1`] = `
<section
  className="row"
>
  <div
    className="grid"
    key="bikeUpload"
  >
    <div
      className="grid-row"
    >
      <div
        className="grid-item--borderless field-label"
      >
        Bike Brand
      </div>
      <div
        className="grid-item--borderless"
      >
        <BrandSelect
          bikeOnly={true}
          brandSelected={1}
          brands={
            Array [
              Object {
                "brand_name": "brand 1",
                "id": 1,
              },
              Object {
                "brand_name": "brand 2",
                "id": 2,
                "supplier": Array [],
              },
              Object {
                "brand_name": "brand 3",
                "id": 3,
                "supplier": Array [
                  1,
                ],
              },
              Object {
                "brand_name": "brand 4",
                "delete": true,
                "id": 4,
              },
              Object {
                "brand_name": "brand 5",
                "changed": true,
                "id": 5,
                "supplier": Array [
                  1,
                  3,
                ],
              },
              Object {
                "brand_name": "brand new",
                "changed": true,
                "dummyKey": "123ABC",
              },
              Object {
                "brand_name": "brand new 2",
                "changed": true,
                "dummyKey": "new",
              },
            ]
          }
          disabled={false}
          error=""
          fieldName="brand"
          isEmptyAllowed={true}
        />
        <button
          onClick={[Function]}
        >
          Add Brand
        </button>
      </div>
    </div>
    <div
      className="grid-row"
    >
      <div
        className="grid-item--borderless field-label"
      >
        Frame Name
      </div>
      <div
        className="grid-item--borderless"
      >
        <FormTextInput
          className=""
          dataType="text"
          error=""
          fieldName="frameName"
          id="frameName"
          inputClass=""
          label=""
          list=""
          placeholder="Frame Name"
          size={100}
          title=""
          value="bicicletta"
        />
      </div>
    </div>
  </div>
</section>
`;

exports[`upload frame details displays when none entered 1`] = `
<section
  className="row"
>
  <div
    className="grid"
    key="bikeUpload"
  >
    <div
      className="grid-row"
    >
      <div
        className="grid-item--borderless field-label"
      >
        Bike Brand
      </div>
      <div
        className="grid-item--borderless"
      >
        <BrandSelect
          bikeOnly={true}
          brands={
            Array [
              Object {
                "brand_name": "brand 1",
                "id": 1,
              },
              Object {
                "brand_name": "brand 2",
                "id": 2,
                "supplier": Array [],
              },
              Object {
                "brand_name": "brand 3",
                "id": 3,
                "supplier": Array [
                  1,
                ],
              },
              Object {
                "brand_name": "brand 4",
                "delete": true,
                "id": 4,
              },
              Object {
                "brand_name": "brand 5",
                "changed": true,
                "id": 5,
                "supplier": Array [
                  1,
                  3,
                ],
              },
              Object {
                "brand_name": "brand new",
                "changed": true,
                "dummyKey": "123ABC",
              },
              Object {
                "brand_name": "brand new 2",
                "changed": true,
                "dummyKey": "new",
              },
            ]
          }
          disabled={false}
          error=""
          fieldName="brand"
          isEmptyAllowed={true}
        />
        <button
          onClick={[Function]}
        >
          Add Brand
        </button>
      </div>
    </div>
    <div
      className="grid-row"
    >
      <div
        className="grid-item--borderless field-label"
      >
        Frame Name
      </div>
      <div
        className="grid-item--borderless"
      >
        <FormTextInput
          className=""
          dataType="text"
          error=""
          fieldName="frameName"
          id="frameName"
          inputClass=""
          label=""
          list=""
          placeholder="Frame Name"
          size={100}
          title=""
          value=""
        />
      </div>
    </div>
  </div>
</section>
`;
